pjsip/src/pjsip-ua/sip_inv.c

```diff
@@ -3970,44 +3970,96 @@ static void inv_handle_incoming_reliable_response(pjsip_inv_session *inv,
+     */
+    if (status != PJ_SUCCESS) {
+	status = pjsip_dlg_create_response(inv->dlg, rdata, PJSIP_SC_OK, NULL, &tdata);
+	goto on_return;
+    }
+
+    /* No SDP in the PRACK request, respond with 200/OK without SDP */
+    if (rdata->msg_info.msg->body == NULL) {
+	status = pjsip_dlg_create_response(inv->dlg, rdata, PJSIP_SC_OK, NULL, &tdata);
+	goto on_return;
     }
 
-    /* If SDP negotiation has been successful, also mark the
-     * SDP negotiation flag in the invite transaction to be
-     * done too.
+    neg_state = pjmedia_sdp_neg_get_state(inv->neg);
+
+    /* If we're waiting for an answer, respond with 200/OK without SDP */
+    if (neg_state == PJMEDIA_SDP_NEG_STATE_LOCAL_OFFER) {
+	status = inv_check_sdp_in_incoming_msg(inv, pjsip_rdata_get_tsx(rdata), rdata);
+	/* If SDP negotiation has been successful, also mark the
+	 * SDP negotiation flag in the invite transaction to be
+	 * done too.
+	 */
+	if (status == PJ_SUCCESS && inv->invite_tsx) {
+	    struct tsx_inv_data *tsx_inv_data;
+
+	    /* Get/attach invite session's transaction data */
+	    tsx_inv_data = (struct tsx_inv_data*)inv->invite_tsx->mod_data[mod_inv.mod.id];
+	    if (tsx_inv_data == NULL) {
+		tsx_inv_data = PJ_POOL_ZALLOC_T(inv->invite_tsx->pool, struct tsx_inv_data);
+		tsx_inv_data->inv = inv;
+		tsx_inv_data->has_sdp = PJ_TRUE;
+		inv->invite_tsx->mod_data[mod_inv.mod.id] = tsx_inv_data;
+	    }
+
+	    tsx_inv_data->sdp_done = PJ_TRUE;
+	}
+	status = pjsip_dlg_create_response(inv->dlg, rdata, PJSIP_SC_OK, NULL, &tdata);
+	goto on_return;
+    }
+
+    /* We receive new offer from remote */
+    inv_check_sdp_in_incoming_msg(inv, pjsip_rdata_get_tsx(rdata), rdata);
+
+    /* Application MUST have supplied the answer by now.
+     * If so, negotiate the SDP.
      */
-    if (status == PJ_SUCCESS && inv->invite_tsx) {
-	struct tsx_inv_data *tsx_inv_data;
+    neg_state = pjmedia_sdp_neg_get_state(inv->neg);
+    if (neg_state != PJMEDIA_SDP_NEG_STATE_WAIT_NEGO ||
+	(status=inv_negotiate_sdp(inv)) != PJ_SUCCESS)
+    {
+	/* Negotiation has failed. If negotiator is still
+	 * stuck at non-DONE state, cancel any ongoing offer.
+	 */
+	neg_state = pjmedia_sdp_neg_get_state(inv->neg);
+	if (neg_state != PJMEDIA_SDP_NEG_STATE_DONE)
+	    pjmedia_sdp_neg_cancel_offer(inv->neg);
 
-	/* Get/attach invite session's transaction data */
-	tsx_inv_data = (struct tsx_inv_data*) 
-		       inv->invite_tsx->mod_data[mod_inv.mod.id];
-	if (tsx_inv_data == NULL) {
-	    tsx_inv_data = PJ_POOL_ZALLOC_T(inv->invite_tsx->pool, 
-					    struct tsx_inv_data);
-	    tsx_inv_data->inv = inv;
-	    tsx_inv_data->has_sdp = PJ_TRUE;
-	    inv->invite_tsx->mod_data[mod_inv.mod.id] = tsx_inv_data;
+	/* The behavior is not fully determined, please check RFC 6337
+	 * Section 2.3 Rejection of an Offer.
+	 * This conflicts with RFC 3262 Section 5
+	 */
+	status = pjsip_dlg_create_response(inv->dlg, rdata, PJSIP_SC_NOT_ACCEPTABLE_HERE, NULL, &tdata);
```

Comment:
Reading both RFCs, I believe the recommended approach is:
`200 PRACK Resp. followed by new offer OR termination of dialog` (table 2 in RFC6337).
So if it's preferable to send 488 Not acceptable, the 488 should be put as the response for the INVITE instead (pattern 5):
`When a UA receives a PRACK request with an offer that it cannot
   accept, it may respond with a 200 response with a syntactically
   correct session description.  Optionally, this may be followed by an
   UPDATE request to rearrange the session parameters if both ends
   support the UPDATE method.  Alternatively, the UA may terminate the
   dialog and send an error response to the INVITE request (Pattern 5).
`

This way it doesn't really conflict with RFC 3262.

Reply:
I can try to implement this, but I need help. How to build "syntactically correct SDP" for the PRACK/200 response if SDP negotiation is failed? If PRACK/200 does not contain SDP answer at all, we immediately got CANCEL from UAC ( there is an example in the original issue #2967 ) And for me it's too complicated, to analyze, if UPDATE can be used, so I would just answer 488 to the INVITE.

Reply:
Yes, 200/PRACK followed by 488/INVITE sounds okay.
I suppose you can use `pjmedia_sdp_neg_get_active_local()`? Alternatively, any dummy SDP should be fine since we're going to reject the call next anyway.

Reply:
I tried to implement this 200/PRACK, 488/INVITE sequence. I used the same code, as for UPDATE request SDP negotiation https://github.com/pjsip/pjproject/blob/85f7a694a82a022860c6f07bbe88342bf619af25/pjsip/src/pjsip-ua/sip_inv.c#L3771 and found, that during `inv_negotiate_sdp()` there is execution of callback `pjsua_call_on_media_update` which can terminate the dialog. So if SDP negotiation failed, pjsua first sends reply to original INVITE, terminating current dialog (415 "Unsupported Media Type"), and only after that it is sending reply 488/UPDATE. I think, that correct behavior must be opposite, 488/UPDATE followed by 415/INVITE. I attached test scripts. What do you think, should I open the new issue with UPDATE processing too, or I can use the same behavior for PRACK?
[uac-upd-sdp-decline_py.txt](https://github.com/pjsip/pjproject/files/8802943/uac-upd-sdp-decline_py.txt)
[uac-upd-sdp-decline_xml.txt](https://github.com/pjsip/pjproject/files/8802945/uac-upd-sdp-decline_xml.txt)

Reply:
Yes, a new issue/PR for the UPDATE processing would be great. Thanks.